core/src/main/java/org/elasticsearch/action/admin/indices/shards/TransportIndicesUnassignedShardsAction.java

```diff
@@ -0,0 +1,186 @@
+            this.listener = listener;
+            this.expectedOps = new CountDown(expectedOps);
+            this.fetchResponses = new ConcurrentLinkedQueue<>();
+        }
+
+        void fetch(Collection<ShardId> shardIds) {
+            for (ShardId shardId : shardIds) {
+                InternalAsyncFetch fetch = new InternalAsyncFetch(logger, "monitor_shard_copies", shardId, listShardsInfo);
+                fetch.fetchData(nodes, metaData, Collections.<String>emptySet());
+            }
+        }
+
+        private class InternalAsyncFetch extends AsyncShardFetch<TransportNodesListGatewayStartedShards.NodeGatewayStartedShards> {
+
+            InternalAsyncFetch(ESLogger logger, String type, ShardId shardId, TransportNodesListGatewayStartedShards action) {
+                super(logger, type, shardId, action);
+            }
+
+            @Override
+            protected synchronized void processAsyncFetch(ShardId shardId, TransportNodesListGatewayStartedShards.NodeGatewayStartedShards[] responses, FailedNodeException[] failures) {
+                fetchResponses.add(new Response(shardId, responses, failures));
+                if (expectedOps.countDown()) {
+                    finish();
+                }
+            }
+
+            void finish() {
+                ImmutableOpenMap.Builder<String, ImmutableOpenIntMap<java.util.List<ShardStatus>>> shardsResponseBuilder = ImmutableOpenMap.builder();
+                ImmutableList.Builder<FailedNodeException> failureBuilder = ImmutableList.builder();
+                while (!fetchResponses.isEmpty()) {
+                    Response res = fetchResponses.remove();
+                    ImmutableOpenIntMap<java.util.List<ShardStatus>> indexShards = shardsResponseBuilder.get(res.shardId.getIndex());
+                    final ImmutableOpenIntMap.Builder<java.util.List<ShardStatus>> indexShardsBuilder;
+                    if (indexShards == null) {
+                        indexShardsBuilder = ImmutableOpenIntMap.builder();
+                    } else {
+                        indexShardsBuilder = ImmutableOpenIntMap.builder(indexShards);
+                    }
+                    java.util.List<ShardStatus> shardStatuses = indexShardsBuilder.get(res.shardId.id());
+                    if (shardStatuses == null) {
+                        shardStatuses = new ArrayList<>();
+                    }
+                    for (TransportNodesListGatewayStartedShards.NodeGatewayStartedShards response : res.responses) {
+                        shardStatuses.add(new ShardStatus(response.getNode().id(), response.version(), response.exception()));
+                    }
+                    indexShardsBuilder.put(res.shardId.id(), shardStatuses);
+                    shardsResponseBuilder.put(res.shardId.getIndex(), indexShardsBuilder.build());
+                    for (FailedNodeException failure : res.failures) {
+                        failureBuilder.add(failure);
```

Comment:
I think we loose the information about which shard has failed. Should we wrap it in DefaultShardOperationFailedException ?